client/comms.c

```diff
@@ -0,0 +1,350 @@
+//-----------------------------------------------------------------------------
+void UsbCommandReceived(UsbCommand *UC)
+{
+	switch(UC->cmd) {
+
+	// First check if we are handling a debug message
+	case CMD_DEBUG_PRINT_STRING: {
+		char s[USB_CMD_DATA_SIZE+1];
+		memset(s, 0x00, sizeof(s));
+		size_t len = MIN(UC->arg[0],USB_CMD_DATA_SIZE);
+		memcpy(s,UC->d.asBytes,len);
+		PrintAndLog("#db# %s", s);
+		return;
+	}
+
+	case CMD_DEBUG_PRINT_INTEGERS: {
+		PrintAndLog("#db# %08x, %08x, %08x       \r\n", UC->arg[0], UC->arg[1], UC->arg[2]);
+		return;
+	}
+
+	case CMD_DOWNLOADED_RAW_ADC_SAMPLES_125K: {
+		// FIXME: This does unsanitised copies into memory when we don't know
+		// the size of the buffer.
+		if (sample_buf) {
+			memcpy(sample_buf+(UC->arg[0]),UC->d.asBytes,UC->arg[1]);
+		}
+		return;
+	}
+
+	default: {
+		storeCommand(UC);
+		return;
+	}
+	}
+}
+
+// Gets a single command from a proxmark3 device. This should never be used
+// with the full client.
+//
+// @param conn A receiver_arg structure.
+// @param command A buffer to store the received command.
+bool ReceiveCommand(receiver_arg* conn, UsbCommand* command) {
+	// Local recieve buffer
+	size_t rxlen;
+	byte_t rx[sizeof(UsbCommand)];
+	byte_t* prx = rx;
+
+	while (conn->run) {
+		rxlen = 0;
+		if (uart_receive(port, prx, sizeof(UsbCommand) - (prx-rx), &rxlen) && rxlen) {
+			prx += rxlen;
+			if (prx-rx < sizeof(UsbCommand)) {
+				// Keep reading until we have a completed response.
+				continue;
+			}
+
+			// We have a completed response.
+			memcpy(command, rx, sizeof(UsbCommand));
+			return true;
+		}
+
+		if (prx == rx) {
+			// We got no complete command while waiting, give up control
+			return false;
+		}
+	}
+
+	// did not get a complete command before being cancelled.
+	return false;
+}
+
+// Worker thread for processing incoming events from the PM3
+void *uart_receiver(void *targ) {
+	receiver_arg *conn = (receiver_arg*)targ;
+	UsbCommand rx;
+
+	while (conn->run) {
+		#ifdef COMMS_DEBUG
+		printf("uart_receiver: get lock\n");
+		#endif
+		// Lock up receives, in case they try to take it away from us.
+		pthread_mutex_lock(&conn->recv_lock);
+		#ifdef COMMS_DEBUG
+		printf("uart_receiver: lock acquired\n");
+		#endif
+
+		if (port == NULL) {
+			#ifdef COMMS_DEBUG
+			printf("uart_receiver: port disappeared\n");
+			#endif
+			// Our port disappeared, stall. This code path matters for the flasher,
+			// where it is fiddling with the serial port under us.
+			pthread_mutex_unlock(&conn->recv_lock);
+			msleep(10);
+			continue;
+		}
+
+		bool got_command = ReceiveCommand(conn, &rx);
+		#ifdef COMMS_DEBUG
+		printf("uart_receiver: got command\n");
+		#endif
+		pthread_mutex_unlock(&conn->recv_lock);
+
+		if (got_command) {
+			UsbCommandReceived(&rx);
+		}
+
+		// We aren't normally trying to transmit in the flasher when the port would
+		// be reset, so we can just keep going at this point.
+		pthread_mutex_lock(&txcmd_lock);
+		if (txcmd_pending) {
```

Comment:
Can't we eliminate txcmd_pending? It is kind of a self made signal.

Reply:
Not entirely.  I'll replace it with an `if` statement instead, because that makes a bit more sense.

The reason is the execution flow is as follows:

1. `SendCommand` called on main thread.
2. `SendCommand` waits for `txcmd_lock`.
3. `SendCommand` checks to see if there is an existing command in the buffer, if not, then stuffs the command in the buffer then returns.
4. `SendCommand` gets another command, gets a lock.
5. `SendCommand` sees there's already a command in the buffer, calls `pthread_cond_wait` to wait for the `txcmd` to be sent, and releases the lock in the interim (and blocks the main thread)
6. `uart_receiver` acquires the `txcmd_lock`
7. `uart_receiver` sees a command in the buffer and sends it.
8. `uart_receiver` clears `txcmd_pending`, and signals to `txcmd_sig` that it is ready for more data
9. `uart_receiver` unlocks `txcmd_lock`
10. `SendCommand` unblocks, locks `txcmd_lock` and stuffs a new command into the buffer.
11. `SendCommand` unlocks `txcmd_lock`, and then returns.
12. `uart_receiver` picks up the new buffered command on the next loop.

It could be replaced also if `txcmd` was a pointer instead, but many operations modify an existing `UsbCommand` immediately after sending it, only to send it again.  This could lead to a race between the main and `uart_receiver` threads, leading to the command possibly being corrupted in flight.

The goal is to make `SendCommand` slightly asynchronous -- it'll only block if there's another command still to be sent out, or if it's a `flasher` command (in which case it is being sent immediately).